File: bridge/web/src/lib/copilot-login.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import type { CopilotLoginStart, CopilotPendingReason } from "./bff-contracts";
import type { DiscoveredModel } from "./types";

export const copilotErrors = {
  copilot_not_ready: "Copilot credential storage is not ready. Ask an operator to repair the workspace grant and enrolled provider store before sign-in. No GitHub token was requested.",
  copilot_storage_unconfirmed: "GitHub approved sign-in, but credential storage could not be confirmed. Ask an operator to check the provider store and grant. Do not repeat approval: the token may already have been consumed and a new sign-in may be required.",
  copilot_invalid_request: "The sign-in request is invalid. Polling stopped.",
  copilot_invalid_response: "The sign-in service returned an invalid response. Polling stopped.",
  copilot_upstream: "GitHub sign-in could not be reached. Polling stopped; check connectivity before trying again.",
  copilot_expired: "The sign-in code expired. A new sign-in is required.",
  copilot_denied: "Sign-in was cancelled on GitHub.",
  copilot_rate_limited: "GitHub rate-limited sign-in. Polling stopped; wait before trying again.",
  copilot_seat_unavailable: "Copilot eligibility could not be verified. Check the account's Copilot seat and Chat access before trying again.",
  transport: "The sign-in connection failed. Polling stopped. Check the connection and provider status before trying again.",
  session: "A signed Bridge operator session is required. Polling stopped.",
} as const;

export type StartResult = { ok: true; data: CopilotLoginStart } | { ok: false; error: string };
export type PollResult =
  | { status: "pending"; interval?: number; reason?: CopilotPendingReason }
  | { status: "authorized"; models: DiscoveredModel[] }
  | { status: "error"; error: string };

function object(value: unknown): Record<string, unknown> | null {
  return value !== null && typeof value === "object" && !Array.isArray(value)
    ? value as Record<string, unknown> : null;
}

function seconds(value: unknown): value is number {
  return typeof value === "number" && Number.isSafeInteger(value) && value >= 1 && value <= 900;
}

export function copilotError(code?: string): string {
  return typeof code === "string" && Object.hasOwn(copilotErrors, code)
    ? copilotErrors[code as keyof typeof copilotErrors] : copilotErrors.transport;
}

function safeMessage(value: unknown): string {
  return Object.values(copilotErrors).some(message => message === value)
    ? value as string : copilotErrors.copilot_invalid_response;
}

export function parseCopilotStart(value: unknown): CopilotLoginStart {
  const data = object(value);
  if (!data || data.error !== undefined || typeof data.device_code !== "string" || !/^[\x21-\x7e]{1,4096}$/.test(data.device_code)
    || typeof data.user_code !== "string" || !/^[A-Z0-9-]{1,64}$/.test(data.user_code)
    || data.verification_uri !== "https://github.com/login/device"
    || !seconds(data.interval) || !seconds(data.expires_in)) {
    throw new Error(copilotErrors.copilot_invalid_response);
  }
  return {
    device_code: data.device_code, user_code: data.user_code, verification_uri: data.verification_uri,
    interval: data.interval, expires_in: data.expires_in,
  };
}

export function parseCopilotPoll(value: unknown): PollResult {
  const data = object(value);
  if (data?.status === "error") return { status: "error", error: safeMessage(data.error) };
  if (data?.error !== undefined) return { status: "error", error: copilotErrors.copilot_invalid_response };
  if (data?.status === "pending"
    && (data.interval === undefined || seconds(data.interval))
    && (data.reason === undefined || data.reason === "authorization_pending" || data.reason === "slow_down")) {
    return { status: "pending", interval: data.interval as number | undefined, reason: data.reason as CopilotPendingReason | undefined };
  }
  if (data?.status === "authorized") {
    // Older BFFs may omit models; other malformed catalogues are not authorization.
    const models = data.models === undefined ? [] : data.models;
    if (Array.isArray(models) && models.length <= 1000) {
      const parsed: DiscoveredModel[] = [];
      for (const value of models) {
        const model = object(value);
        if (!model || typeof model.id !== "string" || !model.id || model.id.length > 512
          || (model.recommended !== undefined && typeof model.recommended !== "boolean")
          || (model.label != null && typeof model.label !== "string")
          || (model.detail != null && typeof model.detail !== "string")) {
          return { status: "error", error: copilotErrors.copilot_invalid_response };
        }
        parsed.push({
          id: model.id, recommended: model.recommended as boolean | undefined,
          label: typeof model.label === "string" ? model.label : null, detail: model.detail as string | undefined,
        });
      }
      return { status: "authorized", models: parsed };
    }
  }
  return { status: "error", error: copilotErrors.copilot_invalid_response };
}

interface Clock {
  now(): number;
  setTimeout(callback: () => void, milliseconds: number): ReturnType<typeof setTimeout>;
  clearTimeout(timer: ReturnType<typeof setTimeout>): void;
}

/** One in-memory attempt; cancellation suppresses callbacks, not an already executing BFF write. */
export function createCopilotLoginController(
  service: { start(): Promise<unknown>; poll(deviceCode: string, interval: number): Promise<unknown> },
  ui: {
    flow(value: CopilotLoginStart | null): void;
    starting(value: boolean): void;
    error(value: string | null): void;
    pending(value: CopilotPendingReason | undefined): void;
    authorized(models: DiscoveredModel[]): void;
  },
  clock: Clock = { now: Date.now, setTimeout, clearTimeout },
) {
  let generation = 0;
  let disposed = false;
  let pollTimer: ReturnType<typeof setTimeout> | undefined;
  let expiryTimer: ReturnType<typeof setTimeout> | undefined;
  function invalidate() {
    generation++;
    if (pollTimer !== undefined) clock.clearTimeout(pollTimer);
    if (expiryTimer !== undefined) clock.clearTimeout(expiryTimer);
    pollTimer = expiryTimer = undefined;
  }
  function clear() {
    ui.flow(null);
    ui.starting(false);
    ui.pending(undefined);
  }
  function fail(message: string) {
    invalidate();
    clear();
    ui.error(message);
  }
  return {
    async begin() {
      if (disposed) return;
      invalidate();
      const attempt = generation;
      const startedAt = clock.now();
      const current = () => !disposed && attempt === generation;
      clear();
      ui.error(null);
      ui.starting(true);
      expiryTimer = clock.setTimeout(() => {
        if (current()) fail(copilotErrors.copilot_expired);
      }, 900_000);
      try {
        const result = object(await service.start());
        if (!current()) return;
        if (result?.ok === false) { fail(safeMessage(result.error)); return; }
        if (result?.ok !== true) { fail(copilotErrors.copilot_invalid_response); return; }
        let flow: CopilotLoginStart;
        try { flow = parseCopilotStart(result.data); }
        catch { fail(copilotErrors.copilot_invalid_response); return; }
        const expiresAt = startedAt + flow.expires_in * 1000;
        if (clock.now() >= expiresAt) { fail(copilotErrors.copilot_expired); return; }
        let interval = flow.interval;
        clock.clearTimeout(expiryTimer!);
        expiryTimer = clock.setTimeout(() => {
          if (current()) fail(copilotErrors.copilot_expired);
        }, expiresAt - clock.now());
        ui.starting(false);
        ui.flow(flow);
        ui.pending(undefined);
        const tick = async () => {
          if (!current()) return;
          if (clock.now() >= expiresAt) { fail(copilotErrors.copilot_expired); return; }
          try {
            const result = parseCopilotPoll(await service.poll(flow.device_code, interval));
            if (!current()) return;
            if (clock.now() >= expiresAt) { fail(copilotErrors.copilot_expired); return; }
            if (result.status === "error") { fail(result.error); return; }
            if (result.status === "authorized") {
              invalidate();
              clear();
              ui.authorized(result.models);
              return;
            }
            interval = Math.max(interval + (result.reason === "slow_down" ? 5 : 0), result.interval ?? interval);
            if (interval > 900) { fail(copilotErrors.copilot_rate_limited); return; }
            ui.pending(result.reason);
            schedule();
          } catch {
            if (current()) fail(copilotErrors.transport);
          }
        };
        const schedule = () => {
          // Schedule after settlement, never on a fixed interval or beyond the original deadline.
          if (clock.now() + interval * 1000 < expiresAt) {
            pollTimer = clock.setTimeout(() => { void tick(); }, interval * 1000);
          }
        };
        schedule();
      } catch {
        if (current()) fail(copilotErrors.transport);
      }
    },
    cancel() {
      if (disposed) return;
      invalidate();
      clear();
      ui.error(null);
    },
    dispose() { disposed = true; invalidate(); },
  };
}

File: bridge/bff/src/routes/operator/copilot_login.rs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

use axum::{
    Json,
    extract::State,
    http::StatusCode,
    response::{IntoResponse, Response},
};
use serde::{Deserialize, Serialize};
use serde_json::{Value, json};

use super::additional_providers::{INFERENCE_PROVIDERS_NS, INFERENCE_PROVIDERS_SECRET};
use super::copilot_transport::{CopilotClient, CopilotEndpoint};
use super::providers::{
    DiscoveredModelDto, copilot_jwt_with_client, fetch_copilot_models_with_client,
    invalidate_copilot_catalog_cache,
};
use crate::{kars::cluster::Cluster, state::AppState};

const CLIENT_ID: &str = "Iv1.b507a08c87ecfe98";
const MAX_SECONDS: u64 = 900;

#[derive(Deserialize, Serialize)]
pub struct CopilotLoginStart {
    pub device_code: String,
    pub user_code: String,
    pub verification_uri: String,
    pub interval: u64,
    pub expires_in: u64,
}

#[derive(Deserialize)]
pub struct CopilotLoginPollRequest {
    pub device_code: String,
    #[serde(default)]
    pub interval: Option<u64>,
}

#[derive(Serialize)]
#[serde(tag = "status", rename_all = "snake_case")]
pub enum CopilotLoginPoll {
    Pending {
        interval: u64,
        reason: PendingReason,
    },
    Authorized {
        models: Vec<DiscoveredModelDto>,
    },
}

#[derive(Serialize)]
#[serde(rename_all = "snake_case")]
pub enum PendingReason {
    AuthorizationPending,
    SlowDown,
}

#[derive(Debug)]
pub enum CopilotLoginError {
    NotReady,
    InvalidRequest,
    Upstream,
    Malformed,
    Expired,
    Denied,
    RateLimited,
    SeatUnavailable,
    StorageUnconfirmed,
}

impl IntoResponse for CopilotLoginError {
    fn into_response(self) -> Response {
        let (status, code, message) = match self {
            Self::NotReady => (
                StatusCode::CONFLICT,
                "copilot_not_ready",
                "Copilot credential storage is not ready. An operator must repair the workspace credential grant and enrolled provider store before sign-in. No GitHub token was requested.",
            ),
            Self::InvalidRequest => (
                StatusCode::BAD_REQUEST,
                "copilot_invalid_request",
                "The sign-in request is invalid.",
            ),
            Self::Upstream => (
                StatusCode::BAD_GATEWAY,
                "copilot_upstream",
                "GitHub sign-in could not be reached. Polling stopped; check connectivity before trying again.",
            ),
            Self::Malformed => (
                StatusCode::BAD_GATEWAY,
                "copilot_invalid_response",
                "GitHub returned an invalid sign-in response. Polling stopped.",
            ),
            Self::Expired => (
                StatusCode::UNPROCESSABLE_ENTITY,
                "copilot_expired",
                "The sign-in code expired. A new sign-in is required.",
            ),
            Self::Denied => (
                StatusCode::UNPROCESSABLE_ENTITY,
                "copilot_denied",
                "Sign-in was cancelled on GitHub.",
            ),
            Self::RateLimited => (
                StatusCode::TOO_MANY_REQUESTS,
                "copilot_rate_limited",
                "GitHub rate-limited sign-in. Polling stopped; wait before trying again.",
            ),
            Self::SeatUnavailable => (
                StatusCode::UNPROCESSABLE_ENTITY,
                "copilot_seat_unavailable",
                "Copilot eligibility could not be verified. Check the account's Copilot seat and Chat access before trying again.",
            ),
            Self::StorageUnconfirmed => (
                StatusCode::CONFLICT,
                "copilot_storage_unconfirmed",
                "GitHub approved sign-in, but credential storage could not be confirmed. An operator must check the provider store and grant. Do not repeat approval: the token may already have been consumed and a new sign-in may be required.",
            ),
        };
        // Only fixed text leaves this endpoint, including logs and unknown upstream errors.
        (
            status,
            Json(json!({"error": {"code": code, "message": message}})),
        )
            .into_response()
    }
}

struct OAuth {
    client: CopilotClient,
}

impl OAuth {
    fn github() -> Result<Self, CopilotLoginError> {
        Ok(Self {
            client: CopilotClient::github().map_err(|_| CopilotLoginError::Upstream)?,
        })
    }

    async fn post(
        &self,
        endpoint: CopilotEndpoint,
        input: Value,
    ) -> Result<Value, CopilotLoginError> {
        let response = self
            .client
            .request(endpoint)
            .header("Accept", "application/json")
            .header("User-Agent", "kars-bridge")
            .json(&input)
            .send()
            .await
            .map_err(|_| CopilotLoginError::Upstream)?;
        if response.status() == StatusCode::TOO_MANY_REQUESTS {
            return Err(CopilotLoginError::RateLimited);
        }
        if !response.status().is_success() {
            return Err(CopilotLoginError::Upstream);
        }
        response
            .json()
            .await
            .map_err(|_| CopilotLoginError::Malformed)
    }
}

fn opaque(value: &str) -> bool {
    !value.is_empty() && value.len() <= 4096 && value.bytes().all(|b| b.is_ascii_graphic())
}

async fn preflight(cluster: &Cluster) -> Result<(), CopilotLoginError> {
    let (_, secret) = cluster
        .integration_store(INFERENCE_PROVIDERS_NS, INFERENCE_PROVIDERS_SECRET)
        .await
        .map_err(|_| CopilotLoginError::NotReady)?;
    if secret
        .metadata
        .resource_version
        .as_deref()
        .is_none_or(str::is_empty)
        || secret
            .data
            .as_ref()
            .is_some_and(|data| data.values().any(|v| std::str::from_utf8(&v.0).is_err()))
    {
        return Err(CopilotLoginError::NotReady);
    }
    Ok(())
}

async fn start(cluster: &Cluster, oauth: &OAuth) -> Result<CopilotLoginStart, CopilotLoginError> {
    preflight(cluster).await?;
    let body = oauth
        .post(
            CopilotEndpoint::DeviceCode,
            json!({"client_id": CLIENT_ID, "scope": "read:user"}),
        )
        .await?;
    if body.get("error").is_some() {
        return Err(CopilotLoginError::Malformed);
    }
    let flow: CopilotLoginStart =
        serde_json::from_value(body).map_err(|_| CopilotLoginError::Malformed)?;
    if !opaque(&flow.device_code)
        || flow.user_code.is_empty()
        || flow.user_code.len() > 64
        || !flow
            .user_code
            .bytes()
            .all(|b| b.is_ascii_uppercase() || b.is_ascii_digit() || b == b'-')
        || flow.verification_uri != "https://github.com/login/device"
        || !(1..=MAX_SECONDS).contains(&flow.interval)
        || !(1..=MAX_SECONDS).contains(&flow.expires_in)
    {
        return Err(CopilotLoginError::Malformed);
    }
    Ok(flow)
}

async fn poll(
    cluster: &Cluster,
    oauth: &OAuth,
    request: CopilotLoginPollRequest,
) -> Result<CopilotLoginPoll, CopilotLoginError> {
    let interval = request.interval.unwrap_or(5);
    if !opaque(&request.device_code) || !(1..=MAX_SECONDS).contains(&interval) {
        return Err(CopilotLoginError::InvalidRequest);
    }
    // Re-check before each exchange; the write still repeats the grant/UID/CAS checks.
    // This is a preflight, not a reservation or a promise that a consumed token can resume.
    preflight(cluster).await?;
    let body = oauth
        .post(
            CopilotEndpoint::AccessToken,
            json!({
                "client_id": CLIENT_ID, "device_code": request.device_code,
                "grant_type": "urn:ietf:params:oauth:grant-type:device_code",
            }),
        )
        .await?;
    if body.get("access_token").is_some() {
        let token = body["access_token"]
            .as_str()
            .filter(|value| opaque(value))
            .ok_or(CopilotLoginError::Malformed)?;
        if body.get("error").is_some()
            || !body["token_type"]
                .as_str()
                .is_some_and(|kind| kind.eq_ignore_ascii_case("bearer"))
        {
            return Err(CopilotLoginError::Malformed);
        }
        let jwt = copilot_jwt_with_client(token, &oauth.client)
            .await
            .map_err(|_| CopilotLoginError::SeatUnavailable)?;
        if !opaque(&jwt) {
            return Err(CopilotLoginError::SeatUnavailable);
        }
        cluster
            .mutate_secret_keys(INFERENCE_PROVIDERS_NS, INFERENCE_PROVIDERS_SECRET, |keys| {
                keys.insert("COPILOT_GITHUB_TOKEN".into(), token.into());
            })
            .await
            .map_err(|_| CopilotLoginError::StorageUnconfirmed)?;
        invalidate_copilot_catalog_cache();
        let models = fetch_copilot_models_with_client(&jwt, &oauth.client)
            .await
            .unwrap_or_default();
        return Ok(CopilotLoginPoll::Authorized { models });
    }
    match body["error"].as_str() {
        Some("authorization_pending") => Ok(CopilotLoginPoll::Pending {
            interval,
            reason: PendingReason::AuthorizationPending,
        }),
        Some("slow_down") => {
            let advertised = match body.get("interval") {
                None => 0,
                Some(value) => value
                    .as_u64()
                    .filter(|v| (1..=MAX_SECONDS).contains(v))
                    .ok_or(CopilotLoginError::Malformed)?,
            };
            let next = interval.saturating_add(5).max(advertised);
            if next > MAX_SECONDS {
                return Err(CopilotLoginError::RateLimited);
            }
            Ok(CopilotLoginPoll::Pending {
                interval: next,
                reason: PendingReason::SlowDown,
            })
        }
        Some("expired_token") => Err(CopilotLoginError::Expired),
        Some("access_denied") => Err(CopilotLoginError::Denied),
        _ => Err(CopilotLoginError::Malformed),
    }
}

pub async fn copilot_login_start(
    State(state): State<AppState>,
) -> Result<Json<CopilotLoginStart>, CopilotLoginError> {
    let cluster = state.cluster().ok_or(CopilotLoginError::NotReady)?;
    start(cluster, &OAuth::github()?).await.map(Json)
}

pub async fn copilot_login_poll(
    State(state): State<AppState>,
    request: Result<Json<CopilotLoginPollRequest>, axum::extract::rejection::JsonRejection>,
) -> Result<Json<CopilotLoginPoll>, CopilotLoginError> {
    let cluster = state.cluster().ok_or(CopilotLoginError::NotReady)?;
    let Json(request) = request.map_err(|_| CopilotLoginError::InvalidRequest)?;
    poll(cluster, &OAuth::github()?, request).await.map(Json)
}

#[cfg(test)]
#[path = "copilot_login_tests.rs"]
mod tests;
